src/Service/RequestForwarder.cpp

```diff
@@ -111,61 +135,55 @@ void RequestForwarder::runReceive(RunnerId runner_id)
+                }
 
-                    client->receive(response);
+                max_wait = std::min(max_wait, static_cast<UInt64>(std::chrono::duration_cast<std::chrono::microseconds>(earliest_request_deadline - now).count()) / 1000);
+            }
+
+            if (!server->isLeader() && server->isLeaderAlive())
+            {
+                int32_t leader = server->getLeader();
+                ptr<ForwardingConnection> connection;
+                {
+                    std::lock_guard<std::mutex> lock(connections_mutex);
+                    connection = connections[leader][runner_id];
+                }
 
-                    if (!response.accepted)
+                if (connection && connection->isConnected())
+                {
+                    if (!connection->poll(max_wait * 1000))
                     {
-                        /// common request
-                        if (response.pkg_type == Data && response.session_id != ForwardResponse::non_session_id)
-                        {
-                            LOG_ERROR(
-                                log,
-                                "Receive failed forward response with type(Result), session {}, xid {}, error code {}",
-                                response.session_id,
-                                response.xid,
-                                response.error_code);
-                            request_processor->onError(
-                                response.accepted,
-                                static_cast<nuraft::cmd_result_code>(response.error_code),
-                                response.session_id,
-                                response.xid,
-                                response.opnum);
-                        }
-                        else if (response.pkg_type == Session)
-                        {
-                            LOG_ERROR(
-                                log,
-                                "Receive failed forward response with type(Session), session {}, xid {}, error code {}",
-                                response.session_id,
-                                response.xid,
-                                response.error_code);
-                        }
-                        else if (response.pkg_type == Handshake)
-                        {
-                            LOG_ERROR(
-                                log,
-                                "Receive failed forward response with type(Handshake), session {}, xid {}, error code {}",
-                                response.session_id,
-                                response.xid,
-                                response.error_code);
-                        }
+                        continue;
                     }
+
+                    ForwardResponsePtr response;
+                    if (connection->receive(response))
+                        processResponse(runner_id, response);
```

Comment:
When `connection->receive` return false? 

Maybe it is better `receive` throw exception instead of return `bool` when something else happens in which fashion the method result is more consistent.